### DISCUSSION TITLE
How to merge custom function with the one provided by LazyVim?

### DISCUSSION THREAD

**Discussion by ericukk** [LazyVim/LazyVim#2357] [Q&A] [answered]
I'd like to have both the behaviour LazyVim provided, and additionally run some custom code after it.
What's the correct way to customise a deep opt that's a function? 
This is what I tried, but got StackOverflow
```
    "hrsh7th/nvim-cmp",
    opts = function(_, opts)
      opts.formatting.format = function(entry, item)
        -- call the fn configured by LazyVim
        local formatted_item = opts.formatting.format(entry, item)
        -- then do more stuff I need ...
        formatted_item.do_more_stuff()
        return formatted_item
      end
    end,
```
I'd like to merge what LazyVim has already configured with my own stuff, similar to decorating a function, or overriding it while calling `super()`.

**Comment by ericukk** [comment]
Kinda related question. If I return a table instead of mutating the opts received in the argument, it doesn't work. It starts throwing errors in other places.
Is it because the merge is only in the first level, meaning that there's no deep merge?
```
-- this didn't work
    opts = function(_, opts)
      return {
        formatting = {
          format = function(_, item)
            local icons = require("lazyvim.config").icons.kinds
            if icons[item.kind] then
              item.kind = icons[item.kind] .. item.kind
            end
            return item
          end,
        },
      }
```

```
   Error  22:53:06 notify.error lazy.nvim Failed to run `config` for nvim-cmp

...l/share/nvim/lazy/LazyVim/lua/lazyvim/plugins/coding.lua:100: bad argument #1 to 'ipairs' (table expected, got nil)

# stacktrace:
  - /LazyVim/lua/lazyvim/plugins/coding.lua:100 _in_ **config**

```

**Reply by dpetka2001** [reply]
No, this error is because you're overwriting the `opts` and now `opts` only contain a table with `formatting = { ...`. But in the `config` function of [nvim-cmp](https://github.com/LazyVim/LazyVim/blob/879e29504d43e9f178d967ecc34d482f902e5a91/lua/lazyvim/plugins/coding.lua#L100) it expects the `opts` to also have a `sources` table (which in your case it doesn't since you're overwriting all `opts` with your own). Thus you get the error in `ipairs` that a `nil` value is being passed instead of a table (no `opts.sources` in your case).

What exactly do you want to do?

**Reply by ericukk** [reply]
I see. So is it correct to say that depending on how you override opts, lazyvim does different things:
- `opts = { some_table }` this will do merge with the default table for 1 level
- `opts = function() return { some_table } end` same as above
- `opts = function(_, opts) ... end` this won't merge, but substitute the original table

What I'm trying to do is have a custom formatting.format function, this fn will call the default one and do something else, kinda like:
```
function my_format()
  super() -- calls the default one
  do_something_else()
end
```

**Reply by dpetka2001** [reply]
No the second one will override the whole default and the third one you use to override specific parts of the default (kind of similar to the first one).

**Reply by dpetka2001** [reply, marked as answer]
You can see in my [dotfiles](https://github.com/dpetka2001/dotfiles/blob/730b614403ceb39a461786c102eb30f51be7c126/dot_config/nvim/lua/plugins/coding.lua#L84-L126) how I do something similar in the `formatting.format` to add some code additional to the default of LazyVim to slightly change the original behavior.

**Reply by ericukk** [reply]
Thanks! How do you find out these subtle of differences?

**Reply by dpetka2001** [reply]
Reading both LazyVim and `lazy.nvim` docs. And whatever other plugin might interest you to make changes. And learn some basic Lua to understand what you're doing.
